tests/test_integrals.py

```diff
@@ -688,3 +718,648 @@ def test_biest_interpolators(self):
+        )
+        np.testing.assert_allclose(data["psi"], psi)
+        np.testing.assert_allclose(data["iota"], iota)
+        assert np.all(data["B^zeta"] > 0)
+        data["Bref"] = 2 * np.abs(psi_boundary) / data["a"] ** 2
+        data["rho"] = rho
+        data["alpha"] = alpha
+        data["zeta"] = zeta
+        data["psi"] = grid.compress(data["psi"])
+        data["iota"] = grid.compress(data["iota"])
+        data["shear"] = grid.compress(data["shear"])
+
+        # Compute analytic approximation.
+        drift_analytic, cvdrift, gbdrift, pitch = TestBounce1D.drift_analytic(data)
+        # Compute numerical result.
+        bounce = Bounce1D(
+            grid.source_grid,
+            data,
+            quad=leggauss(28),  # converges to absolute and relative tolerance of 1e-7
+            Bref=data["Bref"],
+            Lref=data["a"],
+            check=True,
+        )
+        f = Bounce1D.reshape_data(grid.source_grid, cvdrift, gbdrift)
+        drift_numerical_num = bounce.integrate(
+            pitch=pitch[:, np.newaxis],
+            integrand=TestBounce1D.drift_num_integrand,
+            f=f,
+            num_well=1,
+            check=True,
+        )
+        drift_numerical_den = bounce.integrate(
+            pitch=pitch[:, np.newaxis],
+            integrand=TestBounce1D.drift_den_integrand,
+            num_well=1,
+            weight=np.ones(zeta.size),
+            check=True,
+        )
+        drift_numerical = np.squeeze(drift_numerical_num / drift_numerical_den)
+        msg = "There should be one bounce integral per pitch in this example."
+        assert drift_numerical.size == drift_analytic.size, msg
+        np.testing.assert_allclose(
+            drift_numerical, drift_analytic, atol=5e-3, rtol=5e-2
+        )
+
+        self._test_bounce_autodiff(
+            bounce,
+            TestBounce1D.drift_num_integrand,
+            f=f,
+            weight=np.ones(zeta.size),
+        )
+
+        fig, ax = plt.subplots()
+        ax.plot(1 / pitch, drift_analytic)
+        ax.plot(1 / pitch, drift_numerical)
+        return fig
+
+    @staticmethod
+    def _test_bounce_autodiff(bounce, integrand, **kwargs):
+        """Make sure reverse mode AD works correctly on this algorithm."""
+
+        def integrand_grad(*args, **kwargs2):
+            return jnp.vectorize(
+                grad(integrand, -1), signature="()," * len(kwargs["f"]) + "(),()->()"
+            )(*args, *kwargs2.values())
+
+        def fun1(pitch):
+            return jnp.sum(bounce.integrate(pitch, integrand, check=False, **kwargs))
+
+        def fun2(pitch):
+            return jnp.sum(
+                bounce.integrate(pitch, integrand_grad, check=True, **kwargs)
+            )
+
+        pitch = 1.0
+        truth = 650  # Extrapolated from plot.
+        assert np.isclose(grad(fun1)(pitch), truth, rtol=1e-3)
+        # Make sure bounce points get differentiated too.
+        assert np.isclose(fun2(pitch), -131750, rtol=1e-1)
```

Comment:
can you explain a bit what this is testing? Is this like leibniz rule or something?

Reply:
done, will push commits in a batch later